### ISSUE TITLE
Synchronise panel and file system

### ISSUE THREAD

**Issue by FabianSperrle** [getkirby-v2/panel#688]
When discussing #687 @distantnative, @texnixe and I discovered a problem of the panel: When clicking save, the panel just writes its content to a file, thus overriding any changes that might have happened to the file since the panel was loaded. While this might be a rare condition, it does exist and might be an unpleasant surprise for anybody using forms and controllers to store data from the front-end (signup forms, contact forms without mails, ...)

I think the panel should check if the file changed since it was loaded. This might either happen periodically every few seconds, or at least before saving. If the file changed, the content should be loaded anew and the changes made should be applied as a patch. If that is impossible due to conflicts, the user must be made aware of that. 

If there were pre- and post-hooks, the "check the file for changes"-step could be provided by an optional plugin to keep the core simple. 

(edited by @distantnative – just to make the issue even clearer)

**Comment by bastianallgeier** [comment]
I had some kind of lock feature in mind to avoid this. I never ran into it myself, but since you got issues because of it, I will move it further up in the list of additions for a next release.

**Comment by distantnative** [comment]
Since more and more clients have run into this problem (huge thing with comments stored as structure fields), I tried to fig in... and it seems to be rather complicated.

So far I got for the `store.php` the following changes:

```php
public function init($modified) {

    $data       = s::get($this->id());
    $timestamp  = s::get($this->id() . '_time');

    if(!is_array($data) or $timestamp < $modified) {
      $raw = (array)$this->source;
    } else {
      …
    }

    $data = array();

    foreach($raw as $row) {
      …
    }

    …

    if($timestamp < $modified) {
      $this->save();
    }
  }
```

`$modified` gets passed through from `structure` and is `$this->model->modified()`.

It works so far that if the content file gets changed, a reload will display the changes from the file and a warning of "unsaved changes" (which is actually not that correct). Without reload though, there would still be no warning.

**Comment by molocLab** [comment]
Uhm, I miss something....
I modified store.php as described but this error is displayed:

`Warning: Missing argument 1 for Kirby\Panel\Structure\Store::init(), called in... `

**Comment by distantnative** [comment]
@molocLab: You are missing quite a few more changes. The code I posted was just basis for a discussion. It is neither complete nor functioning, so should not use it at all.

**Comment by molocLab** [comment]
@distantnative ok, thanks. I hope this will be fixed for the next release.

**Comment by jenstornell** [comment]
+1 for prioriting on this one, because one of my favorite plugin [Calendar Board](https://github.com/molocLab/kirby-calendar-board/issues/15) is affected by it.

**Comment by molocLab** [comment]
Exactly, my plugin (panel custom field) creates pages and structure fields on the fly, but changes are not displayed, though file system is ok.

I think it would be better to label this issue as "bug".

Thanks.

**Comment by distantnative** [comment]
A first fix has been added to the develop branch: If a content file has changed the panel cache/store will now flushed (so you don't have to sign-out and -in to actually see these changes in the panel). We will think about a way how to deal with changes to the content file between the panel page (re-)load and any form submit.

**Comment by molocLab** [comment]
Fixed for me with 2.3.1 version.

**Comment by lukasbestle** [comment]
2.3.1 didn't fix this, but 2.3.0 might have. :)

**Comment by olach** [comment]
I found this issue from a thread the forum (https://forum.getkirby.com/t/using-hooks-to-write-content-elsewhere-in-site/8448/3?u=ola).

In short: For a site I have some hooks that can update some fields on some pages automatically. This works great, but there can be conflicts if an editor edits a page at the same time a script/hook is updating the content file in the background. When the editor saves the changes, the whole page is overwritten and other changes can be lost.

When reading this thread, it seems to be the same issue as I have. But it does not seem to be fixed when I test it.

Instead of rewriting the whole file upon pressing save in the panel, why not just update/patch the content file with the changed fields? This should be doable:

1. When a page is opened in the panel, all content fields is saved to memory.
2. The editor edits some fields and click save.
3. Now, instead of saving all fields to the content file, Kirby should compare the initial field values, to see what fields the editor has actually changed.
4. The content file is then read, updated with only the changed fields, and then saved to disc.
5. Any fields that was changed by some other hooks/scripts during the time the editor had the page open in the panel should now remain unchanged.

If there is a collision, like a background script had changed the same field as the editor, either the changes made by the editor takes precedence, or a error message is shown when the editor presses save.

Sorry if my comment was unrelated to the initial issue of this thread. But from a quick look it seemed to be the same issue that is not solved what I can see.
